firedrake/matrix_free/preconditioners.py

```diff
@@ -0,0 +1,331 @@
+    velocity.  You must provide these with options using the glbaol
+    option ``Re`` for the Reynolds number and the prefixed option
+    ``pcd_velocity_space`` which should be the index into the full
+    space that gives velocity field.
+
+    """
+    def initialize(self, pc):
+        from firedrake import TrialFunction, TestFunction, dx, \
+            assemble, inner, grad, split, Constant
+        prefix = pc.getOptionsPrefix() + "pcd_"
+
+        # we assume P has things stuffed inside of it
+        _, P = pc.getOperators()
+        context = P.getPythonContext()
+
+        test, trial = context.a.arguments()
+        if test.function_space() != trial.function_space():
+            raise ValueError("Pressure space test and trial space differ")
+
+        Q = test.function_space()
+
+        p = TrialFunction(Q)
+        q = TestFunction(Q)
+
+        mass = p*q*dx
+
+        # Regularisation to avoid having to think about nullspaces.
+        stiffness = inner(grad(p), grad(q))*dx + Constant(1e-6)*p*q*dx
+
+        opts = PETSc.Options()
+        # we're inverting Mp and Kp, so default them to assembled.
+        # Fp only needs its action, so default it to mat-free.
+        # These can of course be overridden.
+        # only Fp is referred to in update, so that's the only
+        # one we stash.
+        Mp_mat_type = opts.getString(prefix+"Mp_mat_type", "")
+        Kp_mat_type = opts.getString(prefix+"Kp_mat_type", "")
+        self.Fp_mat_type = opts.getString(prefix+"Fp_mat_type", "matfree")
+
+        
+        # FIXME: allow these guys to be matrix-free
+        Mp = assemble(mass, form_compiler_parameters=context.fc_params,
+                      mat_type=Mp_mat_type)
+        Kp = assemble(stiffness, form_compiler_parameters=context.fc_params,
+                      mat_type=Kp_mat_type)
+
+        Mp.force_evaluation()
+        Kp.force_evaluation()
+
+        # FIXME: Should we transfer nullspaces over.  I think not.
+
+        Mksp = PETSc.KSP().create()
+        Mksp.setOptionsPrefix(prefix + "Mp_")
+        Mksp.setOperators(Mp.petscmat)
+        Mksp.setUp()
+        Mksp.setFromOptions()
+        self.Mksp = Mksp
+
+        Kksp = PETSc.KSP().create()
+        Kksp.setOptionsPrefix(prefix + "Kp_")
+        Kksp.setOperators(Kp.petscmat)
+        Kksp.setUp()
+        Kksp.setFromOptions()
+        self.Kksp = Kksp
+
+        state = context.context["state"]
+
+        Re = context.context.get("Re", 1.0)
+
+        velid = context.context["velocity_space"]
+
+        u0 = split(state)[velid]
+        fp = 1.0/Re * inner(grad(p), grad(q))*dx + inner(u0, grad(p))*q*dx
+
+        self.Re = Re
+        # FIXME, allow assembled matrix here
+        self.Fp = assemble(fp, form_compiler_parameters=context.fc_params,
+                           mat_type=self.Fp_mat_type)
+        self.Fp.force_evaluation()
+        Fpmat = self.Fp.petscmat
+        self.workspace = [Fpmat.createVecLeft() for i in (0, 1)]
```

Comment:
Why can't we look at all the boundary facets with Dirichlet conditions and
determine u.n on them?  We have an explicit value for the Dirichlet data --
we could compute, say, an integral average like \int_e u.n ds for each
boundary facet e and label the facet as inflow, outflow, or characteristic
based on this value.  Then, we need to enforce Dirichlet BC on some of the
facets and Robin conditions on others (which means setting up a ds measure
on that set of facets).

If Firedrake supports this, we do it at setUp.  Otherwise, we'd need to
pass in some extra context.

On Wed, Aug 17, 2016 at 12:03 PM, Lawrence Mitchell <
notifications@github.com> wrote:

> In firedrake/matrix_free/preconditioners.py
> <https://github.com/firedrakeproject/firedrake/pull/853#discussion_r75163423>
> :
>
> > +        state = context.context["state"]
> > +
> > +        Re = context.context.get("Re", 1.0)
> > +
> > +        velid = context.context["velocity_space"]
> > +
> > +        u0 = split(state)[velid]
> > +        fp = 1.0/Re * inner(grad(p), grad(q))*dx + inner(u0, grad(p))*q*dx
> > +
> > +        self.Re = Re
> > +        # FIXME, allow assembled matrix here
> > +        self.Fp = assemble(fp, form_compiler_parameters=context.fc_params,
> > +                           mat_type=self.Fp_mat_type)
> > +        self.Fp.force_evaluation()
> > +        Fpmat = self.Fp.petscmat
> > +        self.workspace = [Fpmat.createVecLeft() for i in (0, 1)]
>
> So you need to specify a bunch of information (presumably through the
> appctx). Particularly, which subdomains the inflow is applied to and also
> the value that u.n should take. Of course, we don't specify u.n, but rather
> u. As you say, handling all the cases is somewhat tricky. Can we modify
> the existing N-S example to do backward-facing step instead. Then we can
> demonstrate how to do this in the demo?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/firedrakeproject/firedrake/pull/853/files/4ae09350675779089c0b84e43dd28c13a7d4940f#r75163423>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADl6L2XPYTUHxyowjyObK16-gyGOB56iks5qgz7agaJpZM4JiV-g>
> .
>